src/crimson/os/seastore/object_data_handler.h

```diff
@@ -77,8 +118,51 @@ struct ObjectDataBlock : crimson::os::seastore::LogicalCachedExtent {
   }
 
   void logical_on_delta_write() final {
+    if (overwrites) {
+      overwrites->copy_changes(get_bptr());
+      overwrites->clear();
+      overwrites = nullptr;
+    }
     delta.clear();
   }
+
+  bufferptr &get_bptr() override {
+    if (overwrites == nullptr || overwrites->empty()) {
+      return CachedExtent::get_bptr();
+    }
+    ceph::bufferptr b(ceph::buffer::copy(CachedExtent::get_bptr().c_str(), get_length())); 
+    overwrites->copy_changes(b);
```

Comment:
I think we shouldn't apply the pending changes to the shared buffer before committed.

Reply:
Shouldn't we be able to see the recent changes within a transaction? For instance,
TRANSACTION A
Write (A, 4096, 4096)
Read (A, 4096, 4096)
At this point, the result of the Read operation should match what was stored by the Write, right?
These lines create a new bufferptr, copying the pending changes to the bufferptr, and then set the bufferptr to ptr using the set_bptr() method.

Reply:
Ah, it is not modifying the shared bptr, it is copies it out and then apply the pending change.

But why "get_bptr() const" cannot be used? I think `const` is still correct as the class members are unchanged.

Reply:
This is because `set_bptr()` is called to set the `bufferptr `to the `ptr` within  `get_bptr()`. We need a copy of the latest change anyway, so this code creates a copy and assigns it to the `ptr`.

Reply:
I think my concern is about the last step `set_bptr()` -- it is the shared buffer, not the transaction local view.

We shouldn't apply the pending changes to the shared buffer before committed.

Reply:
If really wants to keep and track the copied ptr, I think it should be stored separately, and apply/assign to the original `ptr` upon committing.

Reply:
I see your point now, that's correct.

For `get_bptr() (const)`, probably we can introduce a `overwrite_buf_t::ptr` instead, make it "mutable" so it can be updated regardless of const. Upon commit, apply it to the original ptr with the overwrites. Is that reasonable?

Reply:
I addressed your comment. Can you take a look at the updated commit?